source/util/huffman_codec.h

```diff
@@ -0,0 +1,294 @@
+            out << " " << node->weight;
+        out << " " << code << std::endl;
+      } else {
+        if (node->left)
+          queue.emplace(node->left, code + "0");
+
+        if (node->right)
+          queue.emplace(node->right, code + "1");
+      }
+    }
+  }
+
+  // Returns the Huffman table. The table was built at at construction time,
+  // this function just returns a const reference.
+  const std::unordered_map<Val, std::pair<uint64_t, size_t>>&
+      GetEncodingTable() const {
+    return encoding_table_;
+  }
+
+  // Encodes |val| and stores its Huffman code in the lower |num_bits| of
+  // |bits|. Returns false of |val| is not in the Huffman table.
+  bool Encode(const Val& val, uint64_t* bits, size_t* num_bits) {
+    auto it = encoding_table_.find(val);
+    if (it == encoding_table_.end())
+      return false;
+    *bits = it->second.first;
+    *num_bits = it->second.second;
+    return true;
+  }
+
+  // Reads bits one-by-one using callback |read_bit| until a match is found.
+  // Matching value is stored in |val|. Returns false if |read_bit| terminates
+  // before a code was mathced.
+  // |read_bit| has type bool func(bool* bit). When called, the next bit is
+  // stored in |bit|. |read_bit| returns false if the stream terminates
+  // prematurely.
+  bool DecodeFromStream(const std::function<bool(bool*)>& read_bit, Val* val) {
+    Node* node = root_;
+    while (true) {
+      assert(node);
+
+      if (node->left == nullptr && node->right == nullptr) {
+        *val = node->val;
+        return true;
+      }
+
+      bool go_right;
+      if (!read_bit(&go_right))
+        return false;
+
+      if (go_right)
+        node = node->right;
+      else
+        node = node->left;
+    }
+
+    assert (0);
+    return false;
+  }
+
+ private:
+  // Huffman tree node.
+  struct Node {
+    Val val = Val();
+    uint32_t weight = 0;
+    // Ids are issued sequentially starting from 1. Ids are used as an ordering
+    // tie-breaker, to make sure that the ordering (and resulting coding scheme)
+    // are consistent accross multiple platforms.
+    uint32_t id = 0;
+    Node* left = nullptr;
+    Node* right = nullptr;
+  };
+
+  // Returns true if |left| has bigger weight than |right|. Node ids are
+  // used as tie-breaker.
+  static bool LeftIsBigger(const Node* left, const Node* right) {
+    if (left->weight == right->weight) {
+      assert (left->id != right->id);
+      return left->id > right->id;
+    }
+    return left->weight > right->weight;
+  }
+
+  // Prints subtree (helper function used by PrintTree).
+  static void PrintTreeInternal(std::ostream& out, Node* node, size_t depth) {
+    if (!node)
+      return;
+
+    const size_t kTextFieldWidth = 7;
+
+    if (!node->right && !node->left) {
+      out << node->val << std::endl;
+    } else {
+      if (node->right) {
+        std::stringstream label;
+        label << std::setfill('-') << std::left << std::setw(kTextFieldWidth)
```

Comment:
It's a mutation of a the out parameter.   So you if you want to keep it this way then your method comment should mention this side effect.